File: tests/fixtures/9495/solution.py
from collections import deque


class Solution:
    def solve(self, n: int, board: list) -> int:
        S, T = 0, 1
        N = 2 + n * n
        cap = [[0] * N for _ in range(N)]
        INF = 10 ** 9
        dr = [-1, 1, 0, 0]
        dc = [0, 0, -1, 1]

        initial_empty = 0
        white_count = 0

        for r in range(n):
            for c in range(n):
                ch = board[r][c]
                node = 2 + r * n + c
                if ch == '.':
                    initial_empty += 1
                    cap[node][T] += 1
                elif ch == 'o':
                    white_count += 1
                    cap[S][node] += 1
                    for d in range(4):
                        nr, nc = r + dr[d], c + dc[d]
                        if 0 <= nr < n and 0 <= nc < n and board[nr][nc] == '.':
                            cap[node][2 + nr * n + nc] = INF

        level = [-1] * N
        iter_ = [0] * N

        def bfs() -> bool:
            for i in range(N):
                level[i] = -1
            level[S] = 0
            q = deque([S])
            while q:
                v = q.popleft()
                for u in range(N):
                    if cap[v][u] > 0 and level[u] < 0:
                        level[u] = level[v] + 1
                        q.append(u)
            return level[T] >= 0

        def dfs(v: int, f: int) -> int:
            if v == T:
                return f
            while iter_[v] < N:
                u = iter_[v]
                if cap[v][u] > 0 and level[v] < level[u]:
                    d = dfs(u, min(f, cap[v][u]))
                    if d > 0:
                        cap[v][u] -= d
                        cap[u][v] += d
                        return d
                iter_[v] += 1
            return 0

        flow = 0
        while bfs():
            for i in range(N):
                iter_[i] = 0
            while True:
                d = dfs(S, INF)
                if d == 0:
                    break
                flow += d

        return initial_empty + white_count - flow

File: tests/fixtures/9495/Solution.java
import java.util.*;

public class Solution {
    static final int INF = Integer.MAX_VALUE / 2;

    static class Edge {
        int to, cap, rev;
        Edge(int to, int cap, int rev) { this.to = to; this.cap = cap; this.rev = rev; }
    }

    List<Edge>[] g;
    int[] level, iter;

    @SuppressWarnings("unchecked")
    void initGraph(int n) {
        g = new ArrayList[n];
        for (int i = 0; i < n; i++) g[i] = new ArrayList<>();
        level = new int[n];
        iter = new int[n];
    }

    void addEdge(int from, int to, int cap) {
        g[from].add(new Edge(to, cap, g[to].size()));
        g[to].add(new Edge(from, 0, g[from].size() - 1));
    }

    boolean bfs(int s, int t) {
        Arrays.fill(level, -1);
        Deque<Integer> q = new ArrayDeque<>();
        level[s] = 0;
        q.add(s);
        while (!q.isEmpty()) {
            int v = q.poll();
            for (Edge e : g[v]) {
                if (e.cap > 0 && level[e.to] < 0) {
                    level[e.to] = level[v] + 1;
                    q.add(e.to);
                }
            }
        }
        return level[t] >= 0;
    }

    int dfs(int v, int t, int f) {
        if (v == t) return f;
        for (; iter[v] < g[v].size(); iter[v]++) {
            Edge e = g[v].get(iter[v]);
            if (e.cap > 0 && level[v] < level[e.to]) {
                int d = dfs(e.to, t, Math.min(f, e.cap));
                if (d > 0) {
                    e.cap -= d;
                    g[e.to].get(e.rev).cap += d;
                    return d;
                }
            }
        }
        return 0;
    }

    int maxflow(int s, int t) {
        int flow = 0;
        while (bfs(s, t)) {
            Arrays.fill(iter, 0);
            int d;
            while ((d = dfs(s, t, INF)) > 0) flow += d;
        }
        return flow;
    }

    public int solve(int n, String[] board) {
        int S = 0, T = 1;
        initGraph(2 + n * n);

        int initialEmpty = 0, whiteCount = 0;
        int[] dr = {-1, 1, 0, 0};
        int[] dc = {0, 0, -1, 1};

        for (int r = 0; r < n; r++) {
            for (int c = 0; c < n; c++) {
                char ch = board[r].charAt(c);
                int node = 2 + r * n + c;
                if (ch == '.') {
                    initialEmpty++;
                    addEdge(node, T, 1);
                } else if (ch == 'o') {
                    whiteCount++;
                    addEdge(S, node, 1);
                    for (int d = 0; d < 4; d++) {
                        int nr = r + dr[d], nc = c + dc[d];
                        if (nr >= 0 && nr < n && nc >= 0 && nc < n && board[nr].charAt(nc) == '.') {
                            addEdge(node, 2 + nr * n + nc, INF);
                        }
                    }
                }
            }
        }

        return initialEmpty + whiteCount - maxflow(S, T);
    }
}
